### PR TITLE
[WOR-1364] Create WorkspaceDashboard component and fix wrapWorkspace types

### PR DESCRIPTION
**Pull request by nawatts**
https://broadworkbench.atlassian.net/browse/WOR-1364

The workspace dashboard has proven difficult to write tests for.

Part of that is because the component we actually want to test doesn't exist independently of the WorkspaceContainer wrapper. There is a forwarded ref render function for WorkspaceDashboard, but it's only turned into a component (with forwardRef) in the flow that also wraps it with wrapWorkspace.

This defines WorkspaceDashboard as a component, which can be exported and tested independently of its wrappers.

Removing the Lodash flow from wrapWorkspace revealed that the types for wrapWorkspace are wrong. Currently, they type the wrapper component props the same as the wrapped component props. Actually, the wrapper component takes a workspace namespace and name, while the wrapped component takes those as well as various workspace info loaded by WorkspaceContainer. This also corrects the types for wrapWorkspace.

### CODE REVIEW

File: src/pages/workspaces/workspace/Dashboard/Dashboard.ts [hunk 1]
```diff
@@ -86,25 +92,19 @@ const DashboardAuthContainer = (props: WorkspaceDashboardProps): ReactNode => {
...
   refreshWorkspace: () => void;
   storageDetails: StorageDetails;
   workspace: Workspace;
 }
 
-const WorkspaceDashboardForwardRefRenderFunction = (
-  props: WorkspaceDashboardComponentProps,
-  ref: ForwardedRef<unknown>
-): ReactNode => {
+const WorkspaceDashboard = forwardRef((props: WorkspaceDashboardProps, ref: ForwardedRef<unknown>): ReactNode => {
```
Comment (cahrens) [inline src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: I'm still not clear what the `forwardRef` does for us (why we need it at all). Can you explain?

Comment (cahrens) [reply src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: I guess it is related to forcing a refresh? But who is forcing refreshes of the Dashboard?

File: src/pages/workspaces/workspace/WorkspaceContainer.ts [hunk 2]
```diff
@@ -258,37 +258,39 @@ const WorkspaceAccessError = () => {
   ]);
 };
 
-interface WrapWorkspaceProps {
+export interface WrapWorkspaceOptions {
```
Comment (cahrens) [inline src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: Why are these options instead of props?

File: src/pages/workspaces/workspace/Dashboard/Dashboard.ts (hunk 1 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: `forwardRef` lets WorkspaceDashboard customize the handle it exposes via a ref using `useImperativeHandle`.

https://react.dev/reference/react/useImperativeHandle

Components wrapped by `wrapWorkspace` are expected to have a `refresh` method on their ref handle, which WorkspaceContainer calls when the current tab is clicked.

Comment (cahrens) [reply src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: I'm confused that we have 3 names for what I think always has to be the same thing-- `WorkspaceWrapperProps`, `WorkspaceDashboardProps`, and `DashboardAuthContainerProps`

File: src/pages/workspaces/workspace/WorkspaceContainer.ts (hunk 2 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: For convention. `wrapWorkspace` is not a React component.

File: src/pages/workspaces/workspace/WorkspaceContainer.ts [hunk 3]
```diff
@@ -258,37 +258,39 @@ const WorkspaceAccessError = () => {
...
  * and provides the workspace analysesData and storageDetails
  * */
-export const wrapWorkspace = <T extends WrappedComponentProps>(
-  props: WrapWorkspaceProps
-): WorkspaceWrapperFunction<T> => {
-  const { breadcrumbs, activeTab, title } = props;
-  return (WrappedComponent: WrappedWorkspaceComponent<T>): WrappedWorkspaceComponent<T> => {
-    const Wrapper = (props) => {
+export const wrapWorkspace = (opts: WrapWorkspaceOptions) => {
+  const { breadcrumbs, activeTab, title } = opts;
```
Comment (nawatts) [inline src/pages/workspaces/workspace/WorkspaceContainer.ts]: @cahrens also renaming `props` to `opts` here avoided a shadowed variable. Previously, Wrapper's `props` shadowed the `props` passed to `wrapWorkspace`.

File: src/pages/workspaces/workspace/WorkspaceContainer.ts (hunk 2 above)
Comment (cahrens) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: What is its return type?

File: src/pages/workspaces/workspace/Dashboard/Dashboard.ts (hunk 1 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: The workspace wrapper component and DashboardAuthContainer do take the same props: workspace namespace and name.

WorkspaceDashboard takes those and the additional workspace info loaded by WorkspaceContainer.

I think a little duplication is better than a dependency, so I defined DashboardAuthContainerProps as its own interface, though it is the same as WorkspaceWrapperProps. I think that makes the code easier to read since the props type appears right above the component. Previously, things were defined all out of order so it was very hard to follow. The components are still out of order (DashboardAuthContainer depends on WorkspaceDashboardPage).

I'm also imagining that a natural next step is to split these components out into their own modules.

File: src/pages/workspaces/workspace/WorkspaceContainer.ts (hunk 2 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: `wrapWorkspace` returns a function that takes a component to be wrapped and returns a wrapper component.

```ts
const wrapWorkspace: (opts: WrapWorkspaceOptions) =>
  (WrappedComponent: WrappedWorkspaceComponent) =>
    (props: WorkspaceWrapperProps) => ReactNode
```

Comment (cahrens) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: Is there a way to declare the return type?

Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: Yes, but I thought it was easier to let it be defined implicitly.

File: src/pages/workspaces/workspace/Dashboard/Dashboard.ts (hunk 1 above)
Comment (cahrens) [reply src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: What if we declare an interface just for namespace/name, and have the other ones extend it? What I'm getting at is that I believe all these components need `namespace` and `name`, and changing one of them to not have one or both of those would break things. But right now it seems like it also could just be coincidental that they all  have the same properties.

File: src/pages/workspaces/workspace/WorkspaceContainer.ts (hunk 2 above)
Comment (cahrens) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: Perhaps. When I first looked at it though, to me it looked like a React component (which is a pretty loose definition). Seeing the return type specification may have helped me… though it's possible I wouldn't be able to grok it anyway. :(

Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: 👍 Will add.

File: src/pages/workspaces/workspace/Dashboard/Dashboard.ts (hunk 1 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/Dashboard/Dashboard.ts:107]: I think that would be ok here, but if we split the components into separate modules (as I think we're likely to do), we'd end up either back to duplicating types like this PR or importing types, which I think makes the code more difficult to understand.

File: src/pages/workspaces/workspace/WorkspaceContainer.ts (hunk 2 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceContainer.ts:261]: Added a return type and a comment in https://github.com/DataBiosphere/terra-ui/pull/4507/commits/f064f48e682a75f407d06c90f25896c2abc6f64f.

**Review by cahrens — APPROVED** [review]
Thanks!

**Review by blakery — APPROVED** [review]
